src/Microsoft.OpenApi/Validations/Rules/OpenApiSchemaRules.cs

```diff
@@ -70,15 +68,77 @@ public static class OpenApiSchemaRules
                         }
                     }
 
                     context.Exit();
                 });
+
+        /// <summary>
+        /// Validates the property name in the discriminator against the ones present in the children schema
+        /// </summary>
+        /// <param name="schema">The parent schema.</param>
+        /// <param name="discriminator">Adds support for polymorphism. The discriminator is an object name that is used to differentiate
+        /// between other schemas which may satisfy the payload description.</param>
+        /// <param name="schemaReferenceId"></param>
+        /// <param name="context">A validation context.</param>
+        public static bool ValidateChildSchemaAgainstDiscriminator(OpenApiSchema schema, string discriminator, string schemaReferenceId, IValidationContext context)
+        {
+            bool containsDiscriminator = false;
+
+            if (!schema.Required.Contains(discriminator))
+            {
+                // recursively check nested schema.OneOf, schema.AnyOf or schema.AllOf and their required fields for the discriminator
+                if (schema.OneOf.Count != 0)
+                {
+                    return TraverseSchemaElements(discriminator, schema.OneOf, schemaReferenceId, context, containsDiscriminator);
+                }
+                if (schema.AnyOf.Count != 0)
+                {
+                    return TraverseSchemaElements(discriminator, schema.AnyOf, schemaReferenceId, context, containsDiscriminator);
+                }
+                if (schema.AllOf.Count != 0)
+                {
+                    return TraverseSchemaElements(discriminator, schema.AllOf, schemaReferenceId, context, containsDiscriminator);
+                }
+            }
+
+            return containsDiscriminator;
+        }
+
+        /// <summary>
+        /// Traverses the schema elements and checks whether the schema contains the discriminator.
+        /// </summary>
+        /// <param name="discriminator">Adds support for polymorphism. The discriminator is an object name that is used to differentiate
+        /// between other schemas which may satisfy the payload description.</param>
+        /// <param name="childSchema">The child schema.</param>
+        /// <param name="schemaReferenceId"> The schema reference Id.</param>
+        /// <param name="context"> A validation context.</param>
+        /// <param name="containsDiscriminator">Tracks whether the discriminator is present.</param>
+        /// <returns></returns>
+        public static bool TraverseSchemaElements(string discriminator, IList<OpenApiSchema> childSchema, string schemaReferenceId, IValidationContext context, bool containsDiscriminator)
+        {
+            foreach (var childItem in childSchema)
+            {
+                if (!childItem.Properties.ContainsKey(discriminator) && !childItem.Required.Contains(discriminator))
```

Comment:
```suggestion
                if ((!childItem.Properties?.ContainsKey(discriminatorName) ?? false) &&
                    (!childItem.Required?.Contains(discriminatorName) ?? false))
```
Guard against potential null references.